goalc/compiler/compilation/Type.cpp

```diff
@@ -1003,9 +1003,14 @@ Val* Compiler::compile_heap_new(const goos::Object& form,
       args.push_back(array_size);
     }
 
-    auto array = compile_real_function_call(form, malloc_func, args, env);
-    array->set_type(ts);
-    return array;
+    auto new_array = compile_real_function_call(form, malloc_func, args, env);
+    if (info.is_struct) {
+      new_array->set_type(ts);
+    } else {
+      // allocs of value types are exactly the same as a pointer deref
+      new_array->set_type(TypeSpec("pointer", {ts.get_single_arg()}));
+    }
+    return new_array;
```

Comment:
Another way to do this would be to modify the condition on line 985 from `is_inline` to
```
is_inline && m_ts.lookup_type(elt_type)->is_reference()
```
then I think you can remove the modifications to `get_deref_info`.

I think this would be nicer because it means that we don't add more cases to `get_deref_info`.